source/language/openpulse.rst

```diff
@@ -0,0 +1,697 @@
+  of all future pulses on that frame.
+
+The frame is composed of three parts:
+
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time which is manipulated implicitly and cannot be modified other
+   than through existing timing instructions like ``delay`` and ``barrier``. The time increment
+   is determined by the channel on which the frame is played (see :ref:`Timing` section).
+
+The ``frame`` type is a virtual resource and the exact precision of these parameters is
+hardware specific. It is thus up to the compiler to choose how to implement the required
+transformations to physical resources in hardware (e.g. mapping multiple frames to a
+single NCO).
+
+Frame Construction
+~~~~~~~~~~~~~~~~~~
+
+Frames can be constructed using the ``newframe`` command e.g.
+
+.. code-block:: javascript
+
+  frame driveframe = newframe(5e9, 0.0); // newframe(float[size] frequency, angle[size] phase)
+
+would construct a frame with a frequency of 5 GHz and a phase of 0.0. When
+instantiated, the frame time starts at 0. ``frame``\s can also be copied using the
+``copyframe`` command
+
+.. code-block:: javascript
+
+  frame driveframe1 = newframe(5e9, 0.0);
+  frame driveframe2 = copyframe(driveframe1);
+  driveframe2.phase = driveframe1.phase + pi/2;
+
+This will generate a :math:`\pi/2` phase incremented copy of ``driveframe1`` with the same
+Frame manipulation
+~~~~~~~~~~~~~~~~~~
+
+The (frequency, phase) tuple of a frame can be manipulated throughout program
+by referencing ``.frequency``, and ``.phase``. Operations must be
+appropriate for the respective type, ``float`` for frequency and ``angle`` for
+phase. Again, the exact precision of these calculations is hardware specific.
+
+Here's an example of manipulating the phase to calibrate an ``rz`` gate on a frame called
+``driveframe``:
+
+.. code-block:: javascript
+
+   // Example 1: Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   cal {
+     driveframe.phase += pi/4;
+   }
+
+   // The following is an example only. Frames as arrays has not been agreed on.
+   // This conceptually must be compile-time arrays and treat qubits as indices
+   // which also has not been well-defined. We are exploring other solutions to
+   // the problem of mapping qubits to pulse-level resources.
+
+   // Example 2: Define a calibration for the rz gate on all 8 physical qubits
+   cal {
+     array[frame, 8] rz_frames;
+     frame[0] = newframe(...);
+     // and so on
+   }
+
+   defcal rz(angle[20] theta) $q {
+     rz_frames[q].phase -= theta;
+   }
+
+Manipulating frames based on the state of other frames is also permitted:
+
+.. code-block:: javascript
+
+   // Swap phases between two frames
+   const temp = frame1.phase;
+   frame1.phase = frame2.phase;
+   frame2.phase = temp;
+
+Waveforms
+---------
+
+Waveforms are of type ``waveform`` and can either be:
```

Comment:
Why do we need a `waveform` type instead of using array types like `array[complex[float], len]`?

Reply:
Discussed in meeting, but for the github record two reasons:
1. It is useful for the compiler and control hardware to work with "gaussian" and a few parameters rather than a list of values that happen to be a gaussian
2. Array lengths in OpenQASM are static, so there'd be no way to pass in length as an input